bigquery/google/cloud/bigquery/job.py

```diff
@@ -3007,11 +3025,17 @@ def done(self, retry=DEFAULT_RETRY):
         timeout_ms = None
         if self._done_timeout is not None:
             # Subtract a buffer for context switching, network latency, etc.
-            timeout = self._done_timeout - _TIMEOUT_BUFFER_SECS
-            timeout = max(min(timeout, 10), 0)
-            self._done_timeout -= timeout
+            api_timeout = self._done_timeout - _TIMEOUT_BUFFER_SECS
+            api_timeout = max(min(api_timeout, 10), 0)
+            self._done_timeout -= api_timeout
             self._done_timeout = max(0, self._done_timeout)
-            timeout_ms = int(timeout * 1000)
+            timeout_ms = int(api_timeout * 1000)
+
+        if timeout_ms is not None and timeout_ms > 0:
+            if timeout is not None:
+                timeout = min(timeout_ms / 1000, timeout)
```

Comment:
We want to give ourselves a bit of a buffer (1 second? Whatever we pick, let's make it a private constant) on this timeout. Since `timeout_ms` is the server-side timeout, not the end-to-end timeout, I don't want us to close the connection before we can download the response.

Reply:
Makes sense.

One second (i.e. a small-ish buffer) sounds reasonable. It's enough to make the timeout logic, in principle, correct, as we need to wait for a bit longer than the max server-side processing  time. On the other hand, the buffer should also not be too large, because after all this case is still probably marginal (the server completing the job just before the timeout would kick in).